instance/models/server.py

```diff
@@ -0,0 +1,274 @@
+from instance.models.instance import OpenEdXInstance
+from instance.models.logging_mixin import LoggerMixin
+from instance.models.utils import ValidateModelMixin
+
+
+# Constants ###################################################################
+
+SERVER_STATUS_CHOICES = (
+    ('new', 'New - Not yet loaded'),
+    ('started', 'Started - Running but not active yet'),
+    ('active', 'Active - Running but not booted yet'),
+    ('booted', 'Booted - Booted but not ready to be added to the application'),
+    ('provisioned', 'Provisioned - Provisioning is completed'),
+    ('rebooting', 'Rebooting - Reboot in progress, to apply changes from provisioning'),
+    ('ready', 'Ready - Rebooted and ready to add to the application'),
+    ('live', 'Live - Is actively used in the application and/or accessed by users'),
+    ('stopping', 'Stopping - Stopping temporarily'),
+    ('stopped', 'Stopped - Stopped temporarily'),
+    ('terminating', 'Terminating - Stopping forever'),
+    ('terminated', 'Terminated - Stopped forever'),
+)
+
+
+# Exceptions ##################################################################
+
+class ServerNotReady(Exception):
+    """
+    Raised when an action is attempted in a status that doesn't allow it
+    """
+    pass
+
+
+# Models ######################################################################
+
+class ServerQuerySet(models.QuerySet):
+    """
+    Additional methods for server querysets
+    Also used as the standard manager for the Server model (`Server.objects`)
+    """
+    def terminate(self, *args, **kwargs):
+        """
+        Terminate the servers from the queryset
+        """
+        qs = self.filter(~Q(status='terminated'), *args, **kwargs)
+        for server in qs:
+            server.terminate()
+        return qs
+
+    def exclude_terminated(self):
+        """
+        Filter out terminated servers from the queryset
+        """
+        return self.filter(~Q(status='terminated'))
+
+
+class Server(ValidateModelMixin, TimeStampedModel, LoggerMixin):
+    """
+    A single server VM
+    """
+    status = models.CharField(max_length=11, default='new', choices=SERVER_STATUS_CHOICES, db_index=True)
+
+    objects = ServerQuerySet().as_manager()
+
+    class Meta:
+        abstract = True
+
+    def _set_status(self, status):
+        """
+        Update the current status variable, to be called when a status change is detected
+        """
+        self.status = status
+        self.log('info', 'Changed status for {}: {}'.format(self, self.status))
+        self.save()
+        return self.status
+
+    def sleep_until_status(self, target_status):
+        """
+        Sleep in a loop until the server reaches one of the specified status
+        """
+        self.log('info', 'Waiting for server {} to reach status {}...'.format(self, target_status))
+        while True:
+            self.update_status()
+            if isinstance(target_status, str):
```

Comment:
@antoviaque nit: more concise:

    target_statuses = [target_status] if isinstance(target_status, str) else target_status
    while True:
         self.update_status()
         if self.status in target_statuses:
             break
         time.sleep(1)
    return self.status